Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
Subject: [patch] introduce pause_on_oops_head/tail boot options
From: Ingo Molnar <mingo@elte.hu>

if a system crashes with hard to debug oopses which scroll off the
screen then it's useful to stop the crash right after the register
info or right after the callback printout.

Signed-off-by: Ingo Molnar <mingo@elte.hu>
---
 arch/x86/kernel/traps_32.c |    6 +++++
 arch/x86/kernel/traps_64.c |    2 +
 include/linux/kernel.h     |    4 +++
 kernel/panic.c             |   49 ++++++++++++++++++++++++++++++++++++++++++++-
 4 files changed, 60 insertions(+), 1 deletion(-)

Index: linux-2.6.27-RT/arch/x86/kernel/traps_32.c
===================================================================
--- linux-2.6.27-RT.orig/arch/x86/kernel/traps_32.c
+++ linux-2.6.27-RT/arch/x86/kernel/traps_32.c
@@ -344,9 +344,15 @@ show_stack_log_lvl(struct task_struct *t
 			printk("\n%s       ", log_lvl);
 		printk("%08lx ", *stack++);
 	}
+
+	pause_on_oops_head();
+
 	printk("\n%sCall Trace:\n", log_lvl);
 
 	show_trace_log_lvl(task, regs, sp, bp, log_lvl);
+
+	pause_on_oops_tail();
+
 }
 
 void show_stack(struct task_struct *task, unsigned long *sp)
Index: linux-2.6.27-RT/arch/x86/kernel/traps_64.c
===================================================================
--- linux-2.6.27-RT.orig/arch/x86/kernel/traps_64.c
+++ linux-2.6.27-RT/arch/x86/kernel/traps_64.c
@@ -406,10 +406,12 @@ static void
 show_trace_log_lvl(struct task_struct *task, struct pt_regs *regs,
 		unsigned long *stack, unsigned long bp, char *log_lvl)
 {
+	pause_on_oops_head();
 	printk("\nCall Trace:\n");
 	dump_trace(task, regs, stack, bp, &print_trace_ops, log_lvl);
 	printk("\n");
 	debug_show_held_locks(task);
+	pause_on_oops_tail();
 	print_preempt_trace(task);
 }
 
Index: linux-2.6.27-RT/include/linux/kernel.h
===================================================================
--- linux-2.6.27-RT.orig/include/linux/kernel.h
+++ linux-2.6.27-RT/include/linux/kernel.h
@@ -241,6 +241,10 @@ extern void wake_up_klogd(void);
 extern int oops_in_progress;		/* If set, an oops, panic(), BUG() or die() is in progress */
 extern int panic_timeout;
 extern int panic_on_oops;
+
+extern void pause_on_oops_head(void);
+extern void pause_on_oops_tail(void);
+
 extern int panic_on_unrecovered_nmi;
 extern int tainted;
 extern int unsupported;
Index: linux-2.6.27-RT/kernel/panic.c
===================================================================
--- linux-2.6.27-RT.orig/kernel/panic.c
+++ linux-2.6.27-RT/kernel/panic.c
@@ -31,7 +31,38 @@ static int pause_on_oops;
 static int pause_on_oops_flag;
 static DEFINE_SPINLOCK(pause_on_oops_lock);
 
-int panic_timeout;
+/*
+ * Debugging helper: freeze all console output after printing the
+ * first oops's head (or tail):
+ */
+static int pause_on_oops_head_flag = 0;
+static int pause_on_oops_tail_flag = 0;
+
+static void pause_on_oops_loop(int flag)
+{
+	switch (flag) {
+	default:
+		break;
+	case 1:
+		for (;;)
+			local_irq_disable();
+	case 2:
+		for (;;)
+			local_irq_enable();
+	}
+}
+
+void pause_on_oops_head(void)
+{
+	pause_on_oops_loop(pause_on_oops_head_flag);
+}
+
+void pause_on_oops_tail(void)
+{
+	pause_on_oops_loop(pause_on_oops_tail_flag);
+}
+
+int panic_timeout __read_mostly;
 
 ATOMIC_NOTIFIER_HEAD(panic_notifier_list);
 
@@ -224,6 +255,22 @@ static int __init pause_on_oops_setup(ch
 }
 __setup("pause_on_oops=", pause_on_oops_setup);
 
+static int __init pause_on_oops_head_setup(char *str)
+{
+	pause_on_oops_head_flag = simple_strtoul(str, NULL, 0);
+	printk(KERN_INFO "pause_on_oops_head: %d\n", pause_on_oops_head_flag);
+	return 1;
+}
+__setup("pause_on_oops_head=", pause_on_oops_head_setup);
+
+static int __init pause_on_oops_tail_setup(char *str)
+{
+	pause_on_oops_tail_flag = simple_strtoul(str, NULL, 0);
+	printk(KERN_INFO "pause_on_oops_tail: %d\n", pause_on_oops_tail_flag);
+	return 1;
+}
+__setup("pause_on_oops_tail=", pause_on_oops_tail_setup);
+
 static void spin_msec(int msecs)
 {
 	int i;
